### ISSUE TITLE
PRT can get deleted automatically

### ISSUE THREAD

**Issue by tomaito** [maths/moodle-qtype_stack#131]
Hi,

I just played around with the feedback options in a question. It had 3 inputs with 3 connected PRTs. After removing the commands [[feedback:prt2]] and [[feedback:prt3]] from all the textareas for feedback and hitting save, the complete PRT 2 and 3 vanished. Reinserting the commands did bring back PRTs, but they were blank.

My suggestion would be to either warn the user upon saving a question about this or to keep the the PRTs in the database so they can be restored.

**Comment by sangwinc** [comment]
Yes, I understand this is a frustrating outcome of the design choice and I'm sorry your got caught out by this. We can't keep the PRT in the database unless it is linked to a question.  One design choice would be a "delete PRT" button on the PRT itself, rather than inferring this from the question stem.  I'll ask Tim at the OU if he has a strong view on this change.

**Comment by timhunt** [comment]
I remember we talked about this when building the form.

I think we should try to make it work like this:

* If you delete a PRT reference from the question text or feedback, and then press 'Verify the question text and update the form' the the data should be kept behind the scenes (or even visible on the form) - yes easier to just keep that section of the form visible.
* A warning message should be displayed somewhere (probably near the question text) warning you that you have removed a PRT, and probably with a checkbox to ask if you are sure (but this can be checked by default). (Could also display a warning at the top of that PRT section of the form.)
* When you submit the form. If the checkbox is not checked (or if it is not present because you did not click 'Verify the question text and update the form') then this is treated as a validation error. That will cause the form to be re-displayed with the warning and checkbox.

This should be done the same for Inputs and PRTs.

I may get a chance to code this today. Or, I will not get a chance to do it at all, and someone else will need to.

**Comment by timhunt** [comment]
Right, proposed fix on https://github.com/maths/moodle-qtype_stack/commit/delete-confirm

Would anyone like to test that (both from a no-errors-in-the-code, and a usability point of view) before I shove it into master.

Thanks.

**Comment by tomaito** [comment]
I will test this as soon as I can. Thanks for the reaction on this. :-)

**Comment by sangwinc** [comment]
Thanks for doing this. I've modified the code a little and I've also changed the behaviour so that you must actively check the delete confirmation boxes. On reflection, deleting PRTs and inputs is a very unusual activity, which I'd prefer an active confirmation step of.  Partly because I don't delete them I've barely had this problem!

I've also had some problems with the $fromform field already being populated by a confirm delete field on first load.  If you revert back to a default of 1, then be warned the code will not work.
